File: unidbg-sign/ghidra/decomp.java
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.*;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.symbol.Reference;
import ghidra.util.task.ConsoleTaskMonitor;

public class decomp extends GhidraScript {
    public void run() throws Exception {
        String[] args = getScriptArgs();
        Address base = currentProgram.getImageBase();
        DecompInterface ifc = new DecompInterface();
        ifc.openProgram(currentProgram);
        ConsoleTaskMonitor mon = new ConsoleTaskMonitor();
        for (String off : args) {
            long o = Long.parseLong(off.replace("0x",""),16);
            Address a = base.add(o);
            Function f = getFunctionContaining(a);
            println("==== off="+off+" func="+f+" entry=+0x"+(f!=null?Long.toHexString(f.getEntryPoint().subtract(base)):"?"));
            if (f==null) continue;
            DecompileResults r = ifc.decompileFunction(f, 120, mon);
            if (r!=null && r.getDecompiledFunction()!=null) println(r.getDecompiledFunction().getC());
            for (Function cf : f.getCalledFunctions(mon))
                println("CALL "+cf.getName()+" +0x"+Long.toHexString(cf.getEntryPoint().subtract(base)));
            println("---- XREFS (谁调用本函数) ----");
            for (Reference ref : getReferencesTo(f.getEntryPoint())) {
                Function cf = getFunctionContaining(ref.getFromAddress());
                println("  FROM "+(cf!=null?cf.getName()+" +0x"+Long.toHexString(cf.getEntryPoint().subtract(base)):ref.getFromAddress()));
            }
        }
    }
}

File: unidbg-sign/ghidra/decomp.py
# Ghidra postScript:反编译指定模块偏移的函数 + 列其调用,定位红果签名函数
# args: 偏移列表(hex),如 0x26e684
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

base = currentProgram.getImageBase()
fm = currentProgram.getFunctionManager()
ifc = DecompInterface(); ifc.openProgram(currentProgram)
mon = ConsoleTaskMonitor()

args = getScriptArgs()
offs = args if args else ["0x26e684"]
for off in offs:
    a = base.add(int(off, 16))
    f = fm.getFunctionContaining(a)
    print("==================== off=%s addr=%s func=%s ====================" % (off, a, f))
    if f is None:
        # 强制反汇编+建函数
        print("  (无函数,尝试反汇编)")
        continue
    res = ifc.decompileFunction(f, 90, mon)
    if res and res.getDecompiledFunction():
        c = res.getDecompiledFunction().getC()
        print(c[:6000])
    print("---- callees ----")
    for cf in f.getCalledFunctions(mon):
        print("  CALL %s @ +0x%x" % (cf.getName(), cf.getEntryPoint().subtract(base)))
